### ISSUE TITLE
Impossible to send command to TV

### ISSUE THREAD

**Issue by andreas-bulling** [kdschlosser/samsungctl#40]
Hi,

thanks a lot for this updated library. I am playing around with it to hopefully be able to eventually integrate it with HA.

However, I can't send any commands to my Samsung TV (UE50J6250, firmware 1530). I am using websocket and tried both port 8001 and 8002 but nothing happens. I also don't see the authentication popup which suggests that the connection is actually not established properly.

Anything I can do to debug this / help with to make it easier for you to figure out why it's not working?

**Comment by jnimmo** [comment]
Firstly ensuring you're running the latest version of the code and it isn't running an old version of samsungctl..
Also showing the commands you're trying to use

**Comment by andreas-bulling** [comment]
I am using a fresh checkout from the repository half an hour ago.

I am trying the following:
`samsungctl --host $IP --method websocket --port 8001 KEY_HOME`

**Comment by raydog153** [comment]
Have you validated port 8001 is open on your tv? What is the output?

**Comment by arsaboo** [comment]
Also, shouldn't the command be something like (your are missing the `remoteCmd`) :
`samsungctl --host 192.168.2.252 --method websocket --port 8002 --name remoteCmd KEY_POWER```

**Comment by kdschlosser** [comment]
H and J series TVs use a funky encrypted websocket. At this point in time this library does not support H and J series TV. We are working on getting it added.

If you clone this branch and give it a go. See what happens. 

https://github.com/kdschlosser/samsungctl/tree/H_J_series_TVs?files=1

You need to use 

    --method encrypted

In order to run the proper websocket type for your TV. 

I do not know if that branch will work or not. You are the first person that might have a compatible TV. Let me know if it works or not. If you get any errors please paste the entire output to a comment I need to see all of the information

**Comment by andreas-bulling** [comment]
@kdschlosser I tried the mentioned branch, here is the result:

```
python -m samsungctl --method encrypted --host 192.168.178.60 --interactive                                                     (H_J_series_TVs)
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/andreas/Desktop/samsungctl/samsungctl/__main__.py", line 206, in <module>
    main()
  File "/Users/andreas/Desktop/samsungctl/samsungctl/__main__.py", line 181, in main
    with Remote(config) as remote:
  File "samsungctl/remote.py", line 16, in __init__
    self.remote = RemoteWebsocketEncrypted(config)
  File "samsungctl/remote_encrypted.py", line 73, in __init__
    self.open()
  File "samsungctl/remote_encrypted.py", line 134, in open
    if not self.is_pin_page_open:
  File "samsungctl/remote_encrypted.py", line 118, in is_pin_page_open
    response = requests.get(url)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 72, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 494, in request
    prep = self.prepare_request(req)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 437, in prepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
  File "/usr/local/lib/python2.7/site-packages/requests/models.py", line 305, in prepare
    self.prepare_url(url, params)
  File "/usr/local/lib/python2.7/site-packages/requests/models.py", line 373, in prepare_url
    raise InvalidURL(*e.args)
requests.exceptions.InvalidURL: Failed to parse: 192.168.178.60:None
```

**Comment by kdschlosser** [comment]
-m samsungctl --method encrypted --host $IP --interactive

what is $IP???

it is stating invalid IP is because whatever $IP is it's not the IP address of the TV.

**Comment by andreas-bulling** [comment]
I just replaced the actual IP with a placeholder for posting here. It was an actual IP

**Comment by kdschlosser** [comment]
Oh ok.

**Comment by andreas-bulling** [comment]
Anything to be learned from the output? Can I provide any other help?

**Comment by kdschlosser** [comment]
this is the template being used

    http://{ip}:{port}/ws/apps/CloudPINPage

I am replacing the {ip} with the IP provided by you and 8080 is getting placed where {port} is
so the URL should be correct so long as a proper IP is being supplied.

**Comment by dseifert** [comment]
The final line in the error was "Failed to parse: $IP:None". Assuming @andreas-bulling replaced the actual IP as he said (which is still visible in the beginning of the trace, by the way) and it didn't just get passed in as "$IP", the problem seems to be that port substitution did not work as it said ":None".

**Comment by dseifert** [comment]
@kdschlosser couldn't the issue be in these lines?

```
url = PIN_URL_OPEN.format(
            ip=self.config["host"],
            http_port=self.config["http_port"],
        )
```
I don't see http_port defined anyway, so accessing it should give you the default of `None`.

**Comment by andreas-bulling** [comment]
Yes, I of course had a proper IP address given there and no port. I can try adding a port explicitly but if none is provided the script should IMHO use a default...

**Comment by andreas-bulling** [comment]
Alright, I replaced the http_port entries and tried to connect to port 8001 on my TV. It shows a PIN but when I try to enter it in the console I hit another error:

```
python -m samsungctl --method encrypted --host 192.168.178.60 --port 8001 --interactive                                        (H_J_series_TVs✱)
Enter Pin from TV:3745
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 868, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 741, in format
    return fmt.format(record)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 465, in format
    record.message = record.getMessage()
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 329, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Logged from file remote_encrypted.py, line 99
step 1 response: <!DOCTYPE html>
<html>
<head lang="en">
    <meta charset="UTF-8">
    <title>404 : Not Found</title>

    <link rel="stylesheet" href="/resources/css/bootstrap.min.css">
    <script src="/resources/js/bootstrap.min.js"></script>

</head>
<body>

<div class="container-fluid">
    <div class="page-header">
        <h1>Oops!</h1>
    </div>
    <div class="panel panel-default">
        <div class="panel-heading">404 : Page Not Found</div>
        <div class="panel-body">
            These are not the droids your looking for....
        </div>
    </div>

</div>

</body>
</html>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 868, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 741, in format
    return fmt.format(record)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 465, in format
    record.message = record.getMessage()
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 329, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Logged from file crypto.py, line 59
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 868, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 741, in format
    return fmt.format(record)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 465, in format
    record.message = record.getMessage()
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/logging/__init__.py", line 329, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Logged from file crypto.py, line 62
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/andreas/Desktop/samsungctl/samsungctl/__main__.py", line 206, in <module>
    main()
  File "/Users/andreas/Desktop/samsungctl/samsungctl/__main__.py", line 181, in main
    with Remote(config) as remote:
  File "samsungctl/remote.py", line 16, in __init__
    self.remote = RemoteWebsocketEncrypted(config)
  File "samsungctl/remote_encrypted.py", line 73, in __init__
    self.open()
  File "samsungctl/remote_encrypted.py", line 146, in open
    self.pin = pin
  File "samsungctl/remote_encrypted.py", line 110, in pin
    self.run_forever()
  File "samsungctl/remote_encrypted.py", line 171, in run_forever
    self.pin
  File "samsungctl/pySmartCrypto/crypto.py", line 68, in generate_server_hello
    logger.debug('crypto: aes encrypted: ', encrypted.hex())
AttributeError: 'str' object has no attribute 'hex'
```

**Comment by andreas-bulling** [comment]
I hard-coded a device id (not good if it's empty as it's use during the encryption handshake) and commented out most of the logger calls given that these don't seem to work...

Now I see this:

> Enter Pin from TV:1609
> step 1 response: {"auth_data":""}
> 
> 
> Traceback (most recent call last):
>   File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
>     "__main__", fname, loader, pkg_name)
>   File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
>     exec code in run_globals
>   File "/Users/andreas/Desktop/samsungctl/samsungctl/__main__.py", line 206, in <module>
>     main()
>   File "/Users/andreas/Desktop/samsungctl/samsungctl/__main__.py", line 181, in main
>     with Remote(config) as remote:
>   File "samsungctl/remote.py", line 16, in __init__
>     self.remote = RemoteWebsocketEncrypted(config)
>   File "samsungctl/remote_encrypted.py", line 73, in __init__
>     self.open()
>   File "samsungctl/remote_encrypted.py", line 146, in open
>     self.pin = pin
>   File "samsungctl/remote_encrypted.py", line 110, in pin
>     self.run_forever()
>   File "samsungctl/remote_encrypted.py", line 180, in run_forever
>     GeneratorServerHello=hex(int(server_hello)).upper()
> ValueError: invalid literal for int() with base 10: '\x01\x02'

**Comment by raydog153** [comment]
I hit several of these issues myself and worked thru them. I have a lot more little changes that I needed to make to get to where I am...and am able to pair but not get any farther ATM. 

This is related to #31. Willing to help out any way I can...

**Comment by kdschlosser** [comment]
@andreas-bulling 
No worries m8. I do know how important it is to have a TV integrated into your HA. Thee library in it's current state works for all TV's from 2008 to current except for 2 years. i believe it is 2014 and 2015 the TV's that have H or J in the model number. Some time in 2013 someone decided to see how hard it would be to get the voice controlled TVs to start listening and to intercept that audio. short story is they did it. and they published an article on  Samsung having no kind of security. Samsung went all kinds of bonkers. and they locked down their whole API to everything. 2014 model year devices was the first year to have this new security put into place. it stayed like this for 2 years. When all the static about not having tighter security simmered down they loosened up some. You happen to have one of the TV's made during that 2 years. So it is going to take some to get it working correctly.

@raydog153 
Would you be willing to share the changes you have made with me?

**Comment by raydog153** [comment]
I would be glad to share the changes...had tried to create PR branch but was unable too.

**Comment by kdschlosser** [comment]
zip it up and attach it to a comment.

**Comment by raydog153** [comment]
Here you go...

[samsungctl.zip](https://github.com/kdschlosser/samsungctl/files/2776269/samsungctl.zip)

**Comment by p3g4asus** [comment]
This is my TV (output of http://192.168.25.32:8001/api/v2/):
```json
{
  "id": "068e7780-006e-1000-bc6f-f877b8a47bf1",
  "name": "[TV]Samsung LED55",
  "version": "2.0.24",
  "device": {
    "type": "Samsung SmartTV",
    "duid": "068e7780-006e-1000-bc6f-f877b8a47bf1",
    "model": "14_X14_BT",
    "modelName": "UE55H6400",
    "description": "Samsung TV RCR",
    "networkType": "wireless",
    "ssid": "OpenWrt2",
    "ip": "192.168.25.32",
    "firmwareVersion": "Unknown",
    "name": "[TV]Samsung LED55",
    "id": "068e7780-006e-1000-bc6f-f877b8a47bf1",
    "udn": "068e7780-006e-1000-bc6f-f877b8a47bf1",
    "resolution": "1920x1080",
    "countryCode": "IT",
    "msfVersion": "2.0.24",
    "smartHubAgreement": "true",
    "developerMode": "0",
    "developerIP": "0.0.0.0"
  },
  "type": "Samsung SmartTV",
  "uri": "http://192.168.25.32:8001/api/v2/"
}
```
I made some tests using raydog153 zip with some modifications on main.py to add some command line arguments. This is the detail of the changes made.
After
```python
    parser.add_argument(
        "--port",
        type=int,
        help="TV port number (TCP)"
    )
```
I added

```python
    parser.add_argument(
        "--http_port",
        type=int,
        help="TV port number (TCP)"
    )
    parser.add_argument(
        "--user_id",
        default="0548",
        help="User id"
    )
    parser.add_argument(
        "--device_id",
        default="7890",
        help="Device id"
    )
```

In the TV screen I see two differently shaped windows that show different pin numbers. But the windows are one in front of the other so I can read only the PIN displayed in the front window. This front window disappears when I insert the displayed pin and press enter in the samsungctl console. The other window remains and closes after its countdown reaches zero. I think thet only one of the two windows is supposed to appear and not both.
From the test output it seems I am getting user_id_len = 0
Can you provide the documentation that you are following to implement this comunication protocol so that I can help you in development? I have the hardware to test so it should be much simpler and quicker for me.
Thanks for your work.
The details of the test done follow:
Issued command line: 
```batchfile
python -m samsungctl -vvv --host 192.168.25.32 --method encrypted --http_port 8080 --name samsungctl KEY_MUTE
```
Stdout output:
```
TRACE: request GET URL(200): http://192.168.25.32:8080/ws/apps/CloudPINPage
TRACE: request CONTENT=<?xml version="1.0" encoding="UTF-8"?><service xmlns="urn:dial-multiscreen-org:schemas:dial" xmlns:atom="http://www.w3.org/2005/Atom"><name>CloudPINPage</name><options allowStop="true"/><state>running</state><atom:link rel="run" href="run"/></service>
pairing step 1: http://192.168.25.32:8080/ws/pairing?step=0&app_id=12345&device_id=7890&type=0
TRACE: request GET URL(200): http://192.168.25.32:8080/ws/pairing?step=0&app_id=12345&device_id=7890&type=0
TRACE: request CONTENT={"auth_data":""}


Enter Pin from TV:3620
crypto: pin hash: b65cfa72d8975ff059585739e4959236300da382
crypto: aes: b65cfa72d8975ff059585739e4959236
crypto: aes encrypted: d0c0bb4bb1f7c534d0c1124bf7e3d1239ff109c0b1aae725482221de7fca189b9220eb55658a27fa0a4705325f71ededbee5d919c993495a594d86baaeaafc2881116c02da47847ddbd0cbbc1dd068d2d8795bb1beef4df725d37e6b8fbf72225630c672175c379ae0acd0f82385c8535806df0490aeeb984cd5e65a71e39889
crypto: aes swapped: 00f0a0643b6ce303f1f9eb5239c4d9dd547badced32c04c2763318e2f345281b1731fc293f02f102ced35bce5db2809ccd5e97afd8020d87b2eddc2fda4ec5e6c8c0f2ad5391628e0b49aeced1415201bc9b6431333edaff17ba46ffd27260f0524c3a1517b194a298a4964ed8aae483cfcfdcd785564562610fb851d7e222a9
crypto: data buffer: 000000043035343800f0a0643b6ce303f1f9eb5239c4d9dd547badced32c04c2763318e2f345281b1731fc293f02f102ced35bce5db2809ccd5e97afd8020d87b2eddc2fda4ec5e6c8c0f2ad5391628e0b49aeced1415201bc9b6431333edaff17ba46ffd27260f0524c3a1517b194a298a4964ed8aae483cfcfdcd785564562610fb851d7e222a9
crypto: data hash: 6b954d74f0a22e5934f0c2301644a5dee248a6ac
server hello: 0102000000000000000088000000043035343800f0a0643b6ce303f1f9eb5239c4d9dd547badced32c04c2763318e2f345281b1731fc293f02f102ced35bce5db2809ccd5e97afd8020d87b2eddc2fda4ec5e6c8c0f2ad5391628e0b49aeced1415201bc9b6431333edaff17ba46ffd27260f0524c3a1517b194a298a4964ed8aae483cfcfdcd785564562610fb851d7e222a90000000000
data hash: 6b954d74f0a22e5934f0c2301644a5dee248a6ac
aes key: b65cfa72d8975ff059585739e4959236
pairing step 2: http://192.168.25.32:8080/ws/pairing?step=1&app_id=12345&device_id=7890
TRACE: request POST URL(200): http://192.168.25.32:8080/ws/pairing?step=1&app_id=12345&device_id=7890
TRACE: request PAYLOAD: {"auth_Data": {"auth_type": "SPC", "GeneratorServerHello": "0102000000000000000088000000043035343800F0A0643B6CE303F1F9EB5239C4D9DD547BADCED32C04C2763318E2F345281B1731FC293F02F102CED35BCE5DB2809CCD5E97AFD8020D87B2EDDC2FDA4EC5E6C8C0F2AD5391628E0B49AECED1415201BC9B6431333EDAFF17BA46FFD27260F0524C3A1517B194A298A4964ED8AAE483CFCFDCD785564562610FB851D7E222A90000000000"}}
TRACE: request CONTENT={"auth_data":"{\"auth_type\":\"SPC\",\"request_id\":\"4\",\"GeneratorClientHello\":\"019A000000000000000000\"}"}


step 2 response: {"auth_data":"{\"auth_type\":\"SPC\",\"request_id\":\"4\",\"GeneratorClientHello\":\"019A000000000000000000\"}"}


client hello: 019A000000000000000000
request id: 4
crypto: client hello: 019a000000000000000000
crypto: data_hash: 6b954d74f0a22e5934f0c2301644a5dee248a6ac
crypto: dest: 6b954d74f0a22e5934f0c2301644a5dee248a6ac
crypto: user id:
crypto: pEncWBGx:
crypto: pEncGx:
crypto: pGx:
Traceback (most recent call last):
  File "C:\Python27\lib\runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "C:\Python27\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "D:\Dati\tmp\sam3\samsungctl\__main__.py", line 289, in <module>
    main()
  File "D:\Dati\tmp\sam3\samsungctl\__main__.py", line 257, in main
    with Remote(config, log_level) as remote:
  File "samsungctl\remote.py", line 23, in __init__
    self.remote = RemoteWebsocketEncrypted(config)
  File "samsungctl\remote_encrypted.py", line 74, in __init__
    self.open()
  File "samsungctl\remote_encrypted.py", line 160, in open
    self.pin = pin
  File "samsungctl\remote_encrypted.py", line 115, in pin
    self.run_forever()
  File "samsungctl\remote_encrypted.py", line 232, in run_forever
    self.config["user_id"]
  File "samsungctl\pySmartCrypto\crypto.py", line 139, in parse_client_hello
    bn_pgx = int(codecs.encode(pgx, 'hex'), 16)
ValueError: invalid literal for int() with base 16: ''
```

**Comment by kdschlosser** [comment]
and you also have the same issue that @raydog153 has.. I believe the TV model might be an issue.
UE55H6400

I believe the "Smart" TV models do not start until 7000. I may not be correct in this assessment of what models are going to support the IP control.

I would suggest trying this https://github.com/eclair4151/SmartCrypto and see if it works. What is added in samsungctl is a modified version of that code to be python 2 compatible and also not to be a single use arrangement. 

If that code works then I will go back to step one and add the code in a different manner that would not make so many changes to it.

**Comment by p3g4asus** [comment]
Thank you very much. This evening I will try eclair4151 script. I will report here my results.

**Comment by kdschlosser** [comment]
@p3g4asus 
ok cool. If it works properly then I will add the code with minimal changes to it.

are you familiar with getting the XML files from the TV for UPNP??? I am interested in seeing what the available UPNP functions and classes are. I know that with the legacy TV's (pre 2014) the amount of added control is astounding. I know that Samsung over time as removed bits and pieces of UPNP I would be curious to see what your TV has available to it. The newest TV's offer DLNA via UPNP and a few other functions like setting the aspect ratio. It is a small number of available functions compared to the Legacy TV's. and Since your TV is that middle point I would help me a lot to be able to have the XML files so I can code in better UPNP functionality.

**Comment by raydog153** [comment]
@p3g4asus Since you have port 8001 open then you do not have the same issue as mine, please DO NOT use that zip file I have, it has changes specifically for my model. You could try changing the code where 'type=0' to 'type=1' and that should solve the double pin issue. Also you need to use id/user_id of 12345 else things do not work.

The success logs may definitely help me thou. Your model is very close to mine. I get success but no port 800x to use, and not sure how to re-use the sessionId either. Nor is CloudPINPage ever reported as running for me.

**Comment by kdschlosser** [comment]
OK sweet. it does work on your TV. I will start from the ground up again on the HJ branch. I am going to make less modifications to the code and I am not going to mess about with python 2.7 compatibility just yet.

**Comment by kdschlosser** [comment]
@p3g4asus 

If you take a look at #39 I posted a library I made as well as use directions. I am going to add a feature to it to dump the SSDP packets and the XML files. this way I get a complete layout.

**Comment by raydog153** [comment]
I am using python 2.7 but I think I also have 3 installed, just need to switch my default.

@kdschlosser I think it would be easier to not have that pairing_url method that is dynamic and increments the step. Just call each step explicitly so it is easier to follow the what the code is doing rather than having the code be DRY and more reusable at the moment. I think also better printout of the HTTP request/response with headers would help aid in debugging too.

Example...
```
    API URL GET http://localhost:8080/firstfuel/rest-resource/unauthorize/tenant-info.json?domainName=ut-fe.localhost&tenantId=11111111-1111-1111-1111-111111111111
    API Request Headers: {"User-Agent"=>"Faraday v0.15.3", "accept"=>"application/xml,application/xhtml+xml,application/json,text/html,text/plain", "content-type"=>"application/x-www-form-urlencoded", "FF_PRODUCT"=>"FirstEngage", "FF_REQUEST_ID"=>"72f5c250-3cf4-4a8b-b192-0af7e387173e", "FF_TENANT"=>"11111111-1111-1111-1111-111111111111", "FF_DOMAIN"=>"ut-fe.localhost"}
    API Request Body:
    API Response Headers: {"server"=>"Apache-Coyote/1.1", "cache-control"=>"no-cache, no-store, max-age=0, must-revalidate", "pragma"=>"no-cache", "expires"=>"0", "x-xss-protection"=>"1; mode=block", "x-frame-options"=>"DENY", "x-content-type-options"=>"nosniff", "content-type"=>"application/json;charset=UTF-8", "transfer-encoding"=>"chunked", "date"=>"Sat, 22 Dec 2018 16:00:53 GMT", "connection"=>"close"}
    API Response Body( 582b ): {"tenantId"=>"11111111-1111-1111-1111-111111111111", "tenantName"=>"Unified Tenant", "cmsUrl"=>"http://local-unifiedtenantcms.firstfuelsoftware.net/", "domainSSOSettingsModels"=>[{"id"=>"1fedc7d2-3f34-11e8-ad14-3c9949a356c3", "domainName"=>"ut-fe.localhost", "idpSamlId"=>nil, "idpUserIdType"=>nil, "idpUserIdLocation"=>nil, "idpSsoTargetUrl"=>nil, "idpLogoutUrl"=>nil, "idpLoginUrl"=>nil, "spInitiatedSsoSupported"=>false, "spSamlId"=>nil, "ssoProtocolType"=>nil, "oidcReqFlowType"=>nil, "oidcRespType"=>nil, "oidcClientId"=>nil, "oidcClientSecret"=>nil, "oidcAuthReqUrl"=>nil, "oidcAuthTokenReqUrl"=>nil}]}
    API Completed 200 in 27
```

**Comment by kdschlosser** [comment]
I will mess about with it. I am going to start from the ground up it is going to be the easiest way. The hardest thing is how thee connection was originally coded. I have to get rid of the use of the globals.

**Comment by kdschlosser** [comment]
@p3g4asus i updated that UPNP library I am going to post directions in that issue

**Comment by andreas-bulling** [comment]
I also tried eclair4151's script and it also worked perfectly for my TV:

```
python3 smartcrypto.py                                                                                                       (master✱)
Current state: stopped
Pin NOT on TV
Please enter pin from tv: 1552
Got pin: '1552'

AES key: 440f13f2ffe7800ae87431f50edb70aa
AES encrypted: 9997e28bdf669ac3de57a8275187946ab756aa7bdb45fe39da3111ec91772c84f26765c05e5d1be6824c5911261744c0c1df82c660b691305a54fca47bfc67b0be05446d33a0f3d65b6265377ea921d4ee73753936abc297584cf59d4211d6d3a1e7152c65fcf4acca42f1a34f9fa59ef3c0a63f4b0f32402ae955640199daf6
AES swapped: a256ade07161729ebd899e6adfaf806e5518a220eaaf69cd6964aacfc72c58beffe5bb28fbcf54ab4dd1acd732014a4bc45837eeaececc0bafdc78c7e87d33e903644bad34494693082e302a3db16168debb005a8d5e82e87112fffe12c9c632b231b23ce88fc05c898744e059d365e6fdeeacf5c62f5476f77ca3648806d37d
data buffer: 00000006363534333231A256ADE07161729EBD899E6ADFAF806E5518A220EAAF69CD6964AACFC72C58BEFFE5BB28FBCF54AB4DD1ACD732014A4BC45837EEAECECC0BAFDC78C7E87D33E903644BAD34494693082E302A3DB16168DEBB005A8D5E82E87112FFFE12C9C632B231B23CE88FC05C898744E059D365E6FDEEACF5C62F5476F77CA3648806D37D
hash: 5a41258b84a1237e5ddc5454e258306821782222
secondStepResponse: {"auth_data":"{\"auth_type\":\"SPC\",\"request_id\":\"0\",\"GeneratorClientHello\":\"010100000000000000009E000000063635343332311106C64B8DAFCBD380255FB390D61867B86FF6894B8312ABCD32648375FFF9A23123A32B67BF3E38872B1FB4B60269982D697CF860EC44679565B22438DFF038D41D974B286C7FE2F3E2764EAD86C1C315F12E543598FC0255F326EDD91A35047DCEE168039FD9DE3F017F9C0F0B1E3E9909A50D4B2340D0DA3C29B5B88ACAC3BC204FBE8BE29FC2148E87606608A5995D3043730000000000\"}"}


thirdLen: 138
hello: 010100000000000000009e000000063635343332311106c64b8dafcbd380255fb390d61867b86ff6894b8312abcd32648375fff9a23123a32b67bf3e38872b1fb4b60269982d697cf860ec44679565b22438dff038d41d974b286c7fe2f3e2764ead86c1c315f12e543598fc0255f326edd91a35047dcee168039fd9de3f017f9c0f0b1e3e9909a50d4b2340d0da3c29b5b88acac3bc204fbe8be29fc2148e87606608a5995d3043730000000000
dest: 000000063635343332311106c64b8dafcbd380255fb390d61867b86ff6894b8312abcd32648375fff9a23123a32b67bf3e38872b1fb4b60269982d697cf860ec44679565b22438dff038d41d974b286c7fe2f3e2764ead86c1c315f12e543598fc0255f326edd91a35047dcee168039fd9de3f017f9c0f0b1e3e9909a50d4b2340d0da3c29b5b88acac35a41258b84a1237e5ddc5454e258306821782222
userId: 654321
pEncWBGx: 1106c64b8dafcbd380255fb390d61867b86ff6894b8312abcd32648375fff9a23123a32b67bf3e38872b1fb4b60269982d697cf860ec44679565b22438dff038d41d974b286c7fe2f3e2764ead86c1c315f12e543598fc0255f326edd91a35047dcee168039fd9de3f017f9c0f0b1e3e9909a50d4b2340d0da3c29b5b88acac3
pEncGx: 7587ae632c6dff315a0229514f5869abb8084a49e29e7309ed24eb755482c3def349690fc117cbbd21d82c5374809485f8b0a1ac0c1a3be94e2de8fb366276335d5c597d23d8223d07552bf0b1f3541531d5a1ddccb40c1bb2fdf7b48a87903f2d33d38b468ec0b34df5f65339f89f53c426cfba21a88f359897cb200d1ae9d0
pGx: 8803c97312521aab6b6b5a5c532a39828a27d75f13c16e9f0883318a5fa170e0ae3dcdc3bd3815d23fa92b1195d0c738eff1b00674f3cae02f83a8f6172ea884248af5cfc67cf97ac382de6e001b1bdf077b4187e9f3e8eb097ea746a731034750642397c1c345df80fdde3d470a798bb65b6bce2b6cbda7076c562a05e0b66c
secret: 4f6980af40009502725c676f810fa33b55858552517c43cd1493d75362e22a97401df8b5a3a0df510ef74c7e2e829afe9b53e1cfffa78a0923ff6d8ff56a0ae75cf2d73d8112bf7dcbabd42d39a6ad00d307868fb639ea17f83af76af66692aafbd5ce05d1818aefca7a3c8f31525788f60713464500a33defe511e7a8d679a2
hash2: bc204fbe8be29fc2148e87606608a5995d304373
secret2: 3635343332314f6980af40009502725c676f810fa33b55858552517c43cd1493d75362e22a97401df8b5a3a0df510ef74c7e2e829afe9b53e1cfffa78a0923ff6d8ff56a0ae75cf2d73d8112bf7dcbabd42d39a6ad00d307868fb639ea17f83af76af66692aafbd5ce05d1818aefca7a3c8f31525788f60713464500a33defe511e7a8d679a2
hash3: bc204fbe8be29fc2148e87606608a5995d304373
Pin OK :)

dest_hash: c4410f2605c6d2c96f033143dc9b2c9aec7eb53e
SKPrime: 756804efb9cd9562378aa32bad138c295cdea57f
SKPrimeHash: a04d529e69f2d0cb9fe22f890f0cba6bde71c8b6
ctx: 30f7af3bfc6fcbe601479a25a0800b01
Pin accepted :)

sessionId: 1
SessionID: 1
Authorization successfull :)

Attempting to send command to tv
ws://192.168.178.22:8000/socket.io/1/websocket/fkgPoOKvpa7mEWfyxd4t
Traceback (most recent call last):
  File "smartcrypto.py", line 142, in <module>
    send_command(currentSessionId, ctx, 'KEY_VOLDOWN')
  File "smartcrypto.py", line 106, in send_command
    connection = websocket.create_connection(websocket_url)
AttributeError: module 'websocket' has no attribute 'create_connection'
```

**Comment by kdschlosser** [comment]
OK I am going to work on this for a little bit. see what I can come up with.

**Comment by p3g4asus** [comment]
Attached you will find SSDP.log of my UE55H6400. 
@kdschlosser 
Thank you really much for your hard work
I will investigate and further test to understand what should be saved to shorten the pairing process and send further commands without having to write the pin again. Obviously, having to write the pin each time makes the script pretty unusable for automation / batch purposes. I will report here my findings.


@andreas-bulling 
I am happy that it is working for you too.



[SSDP.log](https://github.com/kdschlosser/samsungctl/files/2780426/SSDP.log)

**Comment by kdschlosser** [comment]
ok so give this whirly bird a twirl.

https://github.com/kdschlosser/samsungctl/tree/H_J

    samsungctl --host 192.168.1.1 --method encrypted

I still have to add the bits so you do not have to enter the pin each time.

**Comment by kdschlosser** [comment]
@p3g4asus

I need the whole folder structure. not just the single file.

**Comment by andreas-bulling** [comment]
> ok so give this whirly bird a twirl.
> 
> https://github.com/kdschlosser/samsungctl/tree/H_J
> 
> ```
> samsungctl --host 192.168.1.1 --method encrypted
> ```
> 
> I still have to add the bits so you do not have to enter the pin each time.

Great - is this based on the latest SmartCrypto version?

**Comment by p3g4asus** [comment]
@kdschlosser 
Apparently only that file was created. I interrupted the scan after 5 minutes of dots.Did I make anything wrong?

**Comment by andreas-bulling** [comment]
`AttributeError: 'RemoteEncrypted' object has no attribute 'tvIP'`

**Comment by p3g4asus** [comment]
@kdschlosser 
Nice work!
It does work. I had only to modify RemoteEncrypted __init__
```python
    def __init__(self, config):
        if 'ctx' in config and config['ctx']:
            self.ctx = config['ctx']
        else:
            self.ctx = None

        if 'session_id' in config and config['session_id']:
            self.currentSessionId = config['session_id']

        self.SKPrime = False
        self.lastRequestId = 0

        self.UserId = "654321"
        self.AppId = "12345"
        self.deviceId = "7e509404-9d7c-46b4-8f6a-e2a9668ad184"
        self.tvIP = config['host']
        self.tvPort = "8080"
        self.aesLib = None
        self.connection = None
        self.config = config
        if not self.ctx:
            self.StartPairing()
```

Then I added 
```python
print("ctx %s " % self.config["ctx"])
print("ses %s " % self.currentSessionId)
```
after
```python
print(websocket_url)
```
I also added two more command line arguments
```python
    parser.add_argument(
        "--ctx",
        default=False,
        help="ctx"
    )
    parser.add_argument(
        "--session_id",
        default=False,
        help="Session ID"
    )
```
First execution command
```batchfile
py -3 -m samsungctl --host 192.168.25.32 --method encrypted KEY_MUTE
```
After the first execution I issued this command for the second execution
```batchfile
py -3 -m samsungctl --host 192.168.25.32 --ctx bb47c08ff166e9d508e4f095475d17e9 --session_id 3 --method encrypted KEY_POWEROFF
```
The values near ctx and session_id were printed after the first execution.
In the second execution I did not have to enter the pin but the command run as expected.

**Comment by kdschlosser** [comment]
all is now fixed.
I have to change out the print statements. other then that it is looking pretty good yes??

**Comment by kdschlosser** [comment]
@p3g4asus 
the UPNP sat there spinning it's wheels?? that is very odd, no errors??? it is defaulted to a timeout of 5 seconds.

**Comment by kdschlosser** [comment]
@p3g4asus 

do me a favor. turn your TV on and on your computer open your web browser to these URL's

http://192.168.25.32:7676/smp_7_
http://192.168.25.32:7676/smp_2_
http://192.168.25.32:7676/smp_15_
http://192.168.25.32:7676/smp_25_

view the page source for each UR L and copy and save the source to separate files. zip the files up and attach them to a post.

**Comment by p3g4asus** [comment]
@kdschlosser 
Today I debugged UPNP_device and found (and solved) 3 issues. I will report here later. Sorry but I fond them at 1.30 am and have no time to answer (too late !! have to sleep :-)) and now it's 8:15 am here in Italy and I am at work.

**Comment by p3g4asus** [comment]
Here I am.
As I sayd I found 3 issues in UPNP_device_develop
1) `sock = send_to(local_address)` in di `discovery.py` `do(local_address, target_ips)` function can throw exceptions if it tryes to send packets to an address that is not in the current subnet range. In my setup this happens several times. So I added a `try...except` block so that the thread is removed from the list as it should be.
2) Path and filename determination is buggy in `upnp_class.py` and `service.py`. it can lead to errors when the line 
```python
path = loc.replace('http://', '').split('/', 1)[-1]
```
generates a path that does not contain '/' (for example when `loc='http://192.168.25.32:7676/smp_7_'`)
I fixed the issue with this snippet:
```python
path = loc.replace('http://', '').split('/', 1)[-1]
if '/' in path:
    path, file_name = path.rsplit('/', 1)
    path = os.path.join(dump, path)
else:
    file_name = path
    path = dump
if not os.path.exists(path):
    os.makedirs(path)
```
3) In Windows file systems in a folder named y you cannot have at the same time a folder named x and a file named x. This with your original library could happen. In my setup it tryed to create inside my dump folder the file `smp_2_\smp_3_` after having created the file `smp_2_` and so it raised a file not found exception because `os.makedirs` fails. So I appended the .xml extension to the files created.
Attached you will find the patched UPNP_device_develop zipped so that it should be easy for you to view the changes using kdiff, winmerge, beyond compare or whatever you prefer.
I will attach also the output of running the test in my setup.
Some empty files are created by `service.py` (don't know if this is correct), while `upnp_class.py` creates not empty files.
No error or exception is thrown now.
Thank you @kdschlosser for your work.
As for the test of the last commit in the H_J branch, I will do it ASAP.
[UPNP_Device.zip](https://github.com/kdschlosser/samsungctl/files/2781887/UPNP_Device.zip)
[output.zip](https://github.com/kdschlosser/samsungctl/files/2781891/output.zip)

**Comment by p3g4asus** [comment]
@kdschlosser 
I see you updated the master branch, so I will test that.
Yesterday when testing the H_J branch I thought that will be geat to have a caching mechanism inside samsungctl for H_J TVs. It should by a set of files inside a cache folder placed inside Appdata or something like that. When one issues a command with --method encrypted you look in the cache folder. If you find inside a file named with deviceId you take ctx and session id from there without asking for pin. Otherwise you make the whole pairing process saving ctx and session paramethers in a json file named with deviceId and placing it in the cache dir. There should be also a command line switch to specify the device id and command line switch to ignore the cache file if present and force to re-do the pairing process.
This is just an idea of improvement that I had. Implement it only if you like and think it can be any good.
Thanks again

**Comment by p3g4asus** [comment]
@kdschlosser 
Another fix for UPNP_device_develop
Now the services file are not empty.
There was another issue.
Attaching
Cheers
[output.zip](https://github.com/kdschlosser/samsungctl/files/2782185/output.zip)
[UPNP_Device.zip](https://github.com/kdschlosser/samsungctl/files/2782186/UPNP_Device.zip)

**Comment by raydog153** [comment]
@p3g4asus I get error with the UPNP code you have....

```
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 163, in _run_module_as_main
    mod_name, _Error)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 102, in _get_module_details
    loader = get_loader(mod_name)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pkgutil.py", line 462, in get_loader
    return find_loader(fullname)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pkgutil.py", line 472, in find_loader
    for importer in iter_importers(fullname):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pkgutil.py", line 428, in iter_importers
    __import__(pkg)
  File "test.py", line 19, in <module>
    for device in UPNP_Device.discover(10, logging.DEBUG, TV_IP_ADDRESS):
  File "/usr/local/lib/python2.7/site-packages/UPNP_Device/__init__.py", line 22, in discover
    yield UPNPObject(addr, locations)
  File "/usr/local/lib/python2.7/site-packages/UPNP_Device/instance_singleton.py", line 11, in __call__
    super(InstanceSingleton, cls).__call__(id, *args)
  File "/usr/local/lib/python2.7/site-packages/UPNP_Device/upnp_class.py", line 44, in __init__
    control_url = service.find('controlURL').text.replace(url, '')
AttributeError: 'NoneType' object has no attribute 'replace'
```

**Comment by p3g4asus** [comment]
@raydog153 
Please try to replace the line of upnp_class.py
```python
control_url = service.find('controlURL').text.replace(url, '')
```
with
```python
try:
    control_url = service.find('controlURL').text.replace(url, '')
except:
    control_url = ''
```

**Comment by raydog153** [comment]
@p3g4asus Thanks, I remember fixing that myself as I hit several of these bugs you fixed when I was trying this out earlier. 

How are you running this thou? I cannot get dump of UPNP data or figure out how....seems it is referencing something in my site-packages. Also I think some of my problems may be due to python version I am using, what version should I be using that is most compatible with current code?

**Comment by raydog153** [comment]
@p3g4asus What firmware is your TV? I have no port 800x working but was hoping that maybe I can enable something in service/debug menu or that I can use some other port open once the bugs are worked out...hence why I am still trying to debug this further.

**Comment by raydog153** [comment]
Thank, that helps, but hangs on me and also looks to be using other IP than the one I entered:
```
rboutotte@MacBook-Pro-2:~/samsungtv/UPNP2/UPNP_Device$ python __main__.py -vvvv --dump ./dump 172.16.16.57
Finding UPNP Devices please wait..
error:ray
local_address: 127.0.0.1

local_address: 127.94.0.1

SSDP: 239.255.255.250
M-SEARCH * HTTP/1.1
ST: upnp:rootdevice
MAN: "ssdp:discover"
HOST: 239.255.255.250:1900
MX: 1
Content-Length: 0


local_address: 127.94.0.2

local_address: 172.16.16.57

SSDP: 239.255.255.250
M-SEARCH * HTTP/1.1
ST: upnp:rootdevice
MAN: "ssdp:discover"
HOST: 239.255.255.250:1900
MX: 1
Content-Length: 0


SSDP: 239.255.255.250
M-SEARCH * HTTP/1.1
ST: upnp:rootdevice
MAN: "ssdp:discover"
HOST: 239.255.255.250:1900
MX: 1
Content-Length: 0


SSDP: 239.255.255.250
M-SEARCH * HTTP/1.1
ST: upnp:rootdevice
MAN: "ssdp:discover"
HOST: 239.255.255.250:1900
MX: 1
Content-Length: 0


SSDP: 239.255.255.250
M-SEARCH * HTTP/1.1
ST: upnp:rootdevice
MAN: "ssdp:discover"
HOST: 239.255.255.250:1900
MX: 1
Content-Length: 0


SSDP: 172.16.16.114 found LOCATION: http://172.16.16.114:8008/ssdp/device-desc.xml
SSDP: 172.16.16.114 found ST: upnp:rootdevice
SSDP: 172.16.16.114 - > HTTP/1.1 200 OK
CACHE-CONTROL: max-age=1800
DATE: Tue, 22 Jan 2019 08:55:46 G1T
EXT:
LOCATION: http://172.16.16.114:8008/ssdp/device-desc.xml
OPT: "http://schemas.upnp.org/upnp/1/0/"; ns=01
01-NLS: 90e87cae-d3af-559d-ad76-d9566abcb5eb
SERVER: QuickPlay/1.0.0.0 UPnP/1.0 AirServer/1.0.0
X-User-Agent: redsonic
ST: upnp:rootdevice
USN: uuid:90e87cae-d3af-559d-ad76-d9566abcb5eb::upnp:rootdevice
BOOTID.UPNP.ORG: 3
CONFIGID.UPNP.ORG: 1


Exception in thread Thread-6:
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/rboutotte/samsungtv/UPNP2/UPNP_Device/UPNP_Device/discover.py", line 224, in found_thread
    found[addr[0]].add(packet['LOCATION'])
KeyError: '172.16.16.114'
```

**Comment by raydog153** [comment]
I think I am using wrong IP, I have several UPNP devices I may have grabbed wrong IP from discovery. Will have to debug further when I get the chance.

**Comment by kdschlosser** [comment]
nope there is still a problem with it. I will take a look and see what changes you mad. You probably only made the changes in a single place when they needed to be made in more then one.

**Comment by kdschlosser** [comment]
Ohhhh you are going to love this. you are going to be able to get things like what the current source is. what thee TV channel is.

when you run the UPNP_device it prints out a whole mess of information. if you redirect that to a file that can help me as well.

    upnp_device --dump "path/to/save/files" > "path/to/save/files/output.txt"

**Comment by p3g4asus** [comment]
I think the problem is on line 218 of discovery.py:
`if 'LOCATION' not in packet:`
should be changed to
`if 'LOCATION' not in packet or addr[0]!=ip:`
Do you agree?

**Comment by p3g4asus** [comment]
@raydog153 
I don't think that is the proper way to solve. I suspect that with your change it will not work properly anymore as it will miss some devices

**Comment by raydog153** [comment]
Thank....I added your other fix and reverted mine and that seems to fix this.
[dump.zip](https://github.com/kdschlosser/samsungctl/files/2783241/dump.zip)

**Comment by p3g4asus** [comment]
From the dump you posted I think that you could control your TV using SDP actions. But @kdschlosser is needed to fully interpret the dump

**Comment by raydog153** [comment]
Yup...hence why I am still trying to debug. I can do screen mirror to my TV, and DLNA works, and has Smart Hub. The wake on lan feature added also works on my TV. The newer firmware's nix'd many of the open ports on these 2014/2015 models ... but since it use to work and I altering the code I can get pin screen to show up I am hopeful I'll this library will be useful.

**Comment by raydog153** [comment]
@kdschlosser Here is dump and output of UN40H5203. I see mention of AllShare, MultiScreen, ScreenMirroringP2PMAC, have been able to control volume thru DLNA.

[dump_UN40H5203.zip](https://github.com/kdschlosser/samsungctl/files/2785118/dump_UN40H5203.zip)
[output.txt](https://github.com/kdschlosser/samsungctl/files/2785119/output.txt)

**Comment by raydog153** [comment]
@kdschlosser This is why I would like all the header info printed. While poking around I see the SDP header indicating application URL when invoking get on 'http://192.168.1.11:7676/smp_22_' which may help aid in debugging...

<img width="441" alt="screen shot 2019-01-22 at 7 28 57 pm" src="https://user-images.githubusercontent.com/233760/51574573-5c9bc080-1e7c-11e9-8d68-47ed0bc61c80.png">

**Comment by p3g4asus** [comment]
@raydog153 @kdschlosser 
Those are some examples of running upnp_client with my TV based on the results of UPNP_device scan output. All the commands are running fine and report correct results.
```batchfile
>pip3 install async-upnp-client
```
```batchfile
>upnp-client --pprint call-action http://192.168.25.32:7676/smp_15_ RC/GetVolume InstanceID=0 Channel=Master
```
```json
{
    "timestamp": 1548203391.915641,
    "service_id": "urn:upnp-org:serviceId:RenderingControl",
    "service_type": "urn:schemas-upnp-org:service:RenderingControl:1",
    "action": "GetVolume",
    "in_parameters": {
        "InstanceID": 0,
        "Channel": "Master"
    },
    "out_parameters": {
        "CurrentVolume": 33
    }
}
```
```batchfile
>upnp-client --pprint call-action http://192.168.25.32:7676/smp_15_ RC/SetMute InstanceID=0 Channel=Master DesiredMute=1
```
```json
{
    "timestamp": 1548203525.0906825,
    "service_id": "urn:upnp-org:serviceId:RenderingControl",
    "service_type": "urn:schemas-upnp-org:service:RenderingControl:1",
    "action": "SetMute",
    "in_parameters": {
        "InstanceID": 0,
        "Channel": "Master",
        "DesiredMute": true
    },
    "out_parameters": {}
}
```
```batchfile
>upnp-client --pprint call-action http://192.168.25.32:7676/smp_15_ RC/SetMute InstanceID=0 Channel=Master DesiredMute=0
```
```json
{
    "timestamp": 1548203543.459175,
    "service_id": "urn:upnp-org:serviceId:RenderingControl",
    "service_type": "urn:schemas-upnp-org:service:RenderingControl:1",
    "action": "SetMute",
    "in_parameters": {
        "InstanceID": 0,
        "Channel": "Master",
        "DesiredMute": false
    },
    "out_parameters": {}
}
```
```btchfile
>upnp-client --pprint call-action http://192.168.25.32:7676/smp_2_ MainTVAgent2/GetCurrentMainTVChannel
```
```json
{
    "timestamp": 1548203756.981378,
    "service_id": "urn:samsung.com:serviceId:MainTVAgent2",
    "service_type": "urn:samsung.com:service:MainTVAgent2:1",
    "action": "GetCurrentMainTVChannel",
    "in_parameters": {},
    "out_parameters": {
        "Result": "OK",
        "CurrentChannel": "<?xml version=\"1.0\" encoding=\"UTF-8\" ?><Channel><ChType>DTV</ChType><MajorCh>1</MajorCh><MinorCh>65534</MinorCh><PTC>32</PTC><ProgNum>3401</ProgNum></Channel>"
    }
}
```

**Comment by kdschlosser** [comment]
OK I updated the UPNP_Device library. I am pretty sure I have the hanging issue solved. I made some changes in the discovery process that I hope will handle any issues.

I also fixed the off SCPDURL that Samsung uses. the lack of the extension specifically. There was also code that was unused that I removed from the EmbeddedDevice class.

There are some difference between how UPNP devices populate the SCPDURL XML node. some will provide the full URL while others will only provide the file name. I can swear I saw a device that provided only the filename and no path to it but a path was needed. I cannot remember what device did that. I am sure I will figure it out if i come across it again. but for the time being I removed the code that was dealing with that because it was causing other issues.

**Comment by kdschlosser** [comment]
I also added the CLI argument --execute

so if you do a printout of a device. You will see spots that state "Access Point" in the printout. you want that value to pass with the --execute.

if you specify -h or --help after --execute it will tell you what you need to pass for parameters.

**Comment by raydog153** [comment]
So the discovery piece no longer hangs. Noticed that I still need to create the folder first prior to dump since root folder name passed in does not get created. Also cannot get the help working or figure out how to pass argument to execute command. Adding -h or --help after --execute yields same exact output.

Attempt to invoke execute:
```python
rboutotte@MacBook-Pro-2:~/samsungtv/UPNP_Device$ python upnp_device --execute UPNPObject.RenderingControl.GetBrightness --help 172.16.16.55
Finding UPNP Devices please wait.......Service location = http://172.16.16.55:7676/smp_4_
.Service location = http://172.16.16.55:7676/smp_22_
Service location = http://172.16.16.55:7676/smp_12_
Service location = http://172.16.16.55:7676/smp_15_
.Service location = http://172.16.16.55:7676/smp_18_

Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/rboutotte/samsungtv/UPNP_Device/upnp_device/__main__.py", line 209, in <module>
    main()
  File "/Users/rboutotte/samsungtv/UPNP_Device/upnp_device/__main__.py", line 156, in main
    type=param.py_data_type,
AttributeError: 'UI4' object has no attribute 'py_data_type'
```

Folder issue:
```python
  File "/usr/local/lib/python2.7/site-packages/UPNP_Device/discover.py", line 101, in convert_ssdp_response
    with open(os.path.join(dump, 'SSDP.log'), 'a') as f:
IOError: [Errno 2] No such file or directory: './dump/SSDP.log'
```

**Comment by kdschlosser** [comment]
I updated it once again. I know for 100% certainty that 100% of the data_type objects now have a data type associated with them. I also now have to make the folder if it does not exist

**Comment by raydog153** [comment]
Hmmm.....then something else is going on. I get same issue, here is log of install for both python 2.7/3 and both failing to run for different reasons.

[console.log](https://github.com/kdschlosser/samsungctl/files/2789765/console.log)

**Comment by kdschlosser** [comment]
for python 2 you are not using the new version of thee library.

the new version does not have print statements that print this out

    Service location = http://192.168.1.11:7676/smp_4_

as far as the python 3 they changed next to __next__ i believe.

**Comment by raydog153** [comment]
So I thought it was something like that, even tried uninstalling it. So I think it may have been due to being on OS with case sensitive file system, anyway past that issue and now it's hanging on me again and same for both versions. Can you tell I'm new to python :)

```python
rboutotte@MacBook-Pro-2:~/samsungtv/UPNP_Device$ python3 upnp_device --execute UPNPObject.RenderingControl.GetBrightness --help 192.168.1.11
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/site-packages/UPNP_Device-0.1.0b0-py3.7.egg/UPNP_Device/discover.py", line 134, in do
    sock = send_to(local_address)
  File "/usr/local/lib/python3.7/site-packages/UPNP_Device-0.1.0b0-py3.7.egg/UPNP_Device/discover.py", line 130, in send_to
    sock.sendto(ssdp_packet.encode('utf-8'), (destination, 1900))
OSError: [Errno 49] Can't assign requested address

Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/site-packages/UPNP_Device-0.1.0b0-py3.7.egg/UPNP_Device/discover.py", line 134, in do
    sock = send_to(local_address)
  File "/usr/local/lib/python3.7/site-packages/UPNP_Device-0.1.0b0-py3.7.egg/UPNP_Device/discover.py", line 130, in send_to
    sock.sendto(ssdp_packet.encode('utf-8'), (destination, 1900))
OSError: [Errno 49] Can't assign requested address
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/site-packages/UPNP_Device-0.1.0b0-py3.7.egg/UPNP_Device/discover.py", line 134, in do
    sock = send_to(local_address)
  File "/usr/local/lib/python3.7/site-packages/UPNP_Device-0.1.0b0-py3.7.egg/UPNP_Device/discover.py", line 130, in send_to
    sock.sendto(ssdp_packet.encode('utf-8'), (destination, 1900))
OSError: [Errno 49] Can't assign requested address


^CTraceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "upnp_device/__main__.py", line 209, in <module>
    main()
  File "upnp_device/__main__.py", line 100, in main
    event.wait(1)
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 552, in wait
    signaled = self._cond.wait(timeout)
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 300, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt
Finding UPNP Devices please wait......
rboutotte@MacBook-Pro-2:~/samsungtv/UPNP_Device$
```

**Comment by raydog153** [comment]
So I forget how I fixed this before, but it has something to do with the adapter_ips. Here is my adapter_ips: ['127.0.0.1', '127.94.0.1', '127.94.0.2', '192.168.1.42']

It should only use the non 127.x.x.x.

**Comment by raydog153** [comment]
Success at last...at least for getting the help command to work for execute:
```python
rboutotte@MacBook-Pro-2:~/samsungtv/UPNP_Device$ python3 upnp_device --execute UPNPObject.RenderingControl.GetBrightness --help 192.168.1.11
Finding UPNP Devices please wait..........
usage: --execute RenderingControl.GetBrightness [-h] --InstanceID INSTANCEID

optional arguments:
  -h, --help            show this help message and exit
  --InstanceID INSTANCEID
rboutotte@MacBook-Pro-2:~/samsungtv/UPNP_Device$
```

For now changed line 59 in discover.py to:
```python
            if isinstance(adapter_ip.ip, tuple) or adapter_ip.nice_name == 'lo0':
```

**Comment by raydog153** [comment]
As well as success for executing:
```
rboutotte@MacBook-Pro-2:~/samsungtv/UPNP_Device$ python3 upnp_device --execute UPNPObject.RenderingControl.GetBrightness --InstanceID 0 192.168.1.11
Finding UPNP Devices please wait..........
CurrentBrightness: 32
```

This is great, gives me a starting point for getting some things working. Going to explore this a bit more.

This issue is also gotten too long, and wrong topic as we are in samsungctl but fixing issue under UPNP_Device repo. Should we get this closed (is original issue fixed)?

**Comment by kdschlosser** [comment]
ok cool i will add that fix. open an issue in the UPNP_Device repo.

**Comment by kdschlosser** [comment]
OK so this should be all set at this point. the original issue was not being able to connect with an encrypted websocket. and that has now been added to the library.
